File: src/main/java/net/hectus/neobb/game/Game.java
package net.hectus.neobb.game;

import com.marcpg.libpg.data.time.Time;
import com.marcpg.libpg.lang.Translation;
import com.marcpg.libpg.util.Randomizer;
import net.hectus.neobb.NeoBB;
import net.hectus.neobb.Rating;
import net.hectus.neobb.cosmetic.EffectManager;
import net.hectus.neobb.event.custom.CancellableImpl;
import net.hectus.neobb.game.util.Difficulty;
import net.hectus.neobb.game.util.*;
import net.hectus.neobb.player.NeoPlayer;
import net.hectus.neobb.player.TargetObj;
import net.hectus.neobb.turn.DummyTurn;
import net.hectus.neobb.turn.Turn;
import net.hectus.neobb.turn.default_game.TTimeLimit;
import net.hectus.neobb.turn.default_game.attributes.clazz.Clazz;
import net.hectus.neobb.turn.default_game.warp.TDefaultWarp;
import net.hectus.neobb.turn.default_game.warp.WarpTurn;
import net.hectus.neobb.turn.person_game.categorization.WinConCategory;
import net.hectus.neobb.util.*;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.title.Title;
import org.bukkit.*;
import org.bukkit.entity.Player;
import org.bukkit.event.Cancellable;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public abstract class Game extends Modifiers.Modifiable {
    public final String id = Randomizer.generateRandomString(10, "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz");
    public final Arena arena;
    public final EffectManager effectManager = new EffectManager();
    public final TurnScheduler turnScheduler = new TurnScheduler();
    public final Difficulty difficulty;

    protected final List<NeoPlayer> initialPlayers = new ArrayList<>();
    protected final List<NeoPlayer> players = new ArrayList<>();
    protected final List<Turn<?>> history = new ArrayList<>();
    protected final Set<Class<? extends Clazz>> allowedClazzes = new HashSet<>();

    protected boolean started;
    protected int turningIndex = 0;

    protected WarpTurn warp;
    protected MinecraftTime time;

    protected Time timeLeft;
    protected int turnCountdown;

    protected Game(World world, Player player) {
        this.difficulty = Difficulty.NORMAL;
        this.arena = new Arena(this, world);
        this.warp = new TDefaultWarp(world);
        resetTurnCountdown();

        NeoPlayer p = new NeoPlayer(player, this);
        initialPlayers.add(p);
        players.add(p);
    }

    protected Game(Difficulty difficulty, World world, @NotNull List<Player> players, WarpTurn defaultWarp) {
        this.difficulty = difficulty;
        this.arena = new Arena(this, world);
        this.warp = defaultWarp;
        this.allowedClazzes.addAll(warp.allows());
        resetTurnCountdown();

        players.forEach(p -> {
            cleanPlayer(p);
            p.setGameMode(GameMode.SURVIVAL);
        });

        double radius = 2.5;
        int playerCount = players.size();

        for (int i = 0; i < playerCount; i++) {
            double angle = 2 * Math.PI * i / playerCount;
            players.get(i).teleport(new Location(warp.center.getWorld(),
                    warp.center.x() + radius * Math.cos(angle),
                    warp.center.y(),
                    warp.center.z() + radius * Math.sin(angle),
                    (float) Math.toDegrees(angle + Math.PI / 2), 0 // No, I am not a math guy...
            ));
        }

        players.stream()
                .map(player -> new NeoPlayer(player, this))
                .peek(initialPlayers::add)
                .forEach(this.players::add);

        NeoBB.LOG.info("{}: Initialized the game.", id);
        GameManager.add(this);
    }

    public abstract GameInfo info();

    public final void turn(@NotNull Turn<?> turn, Cancellable event) {
        NeoPlayer player = turn.player();

        if (outOfBounds(turn.location(), event)) {
            NeoBB.LOG.info("{}: {} used {} out of bounds.", id, player.player.getName(), turn.getClass().getName().replace("net.hectus.neobb.turn.", ""));
            event.setCancelled(true);
            return;
        }

        if (turn instanceof DummyTurn) {
            history.add(turn);
            return;
        }

        if (difficulty.usageRules && turn.unusable()) {
            NeoBB.LOG.info("{}: {} used {} incorrectly.", id, player.player.getName(), turn.getClass().getName().replace("net.hectus.neobb.turn.", ""));
            player.sendMessage(Component.text("That is not quite how to use this turn...", Colors.NEGATIVE));
            player.player.playSound(player.player, Sound.ENTITY_VILLAGER_NO, 1.0f, 1.0f);
            nextTurn(turn);
        }

        boolean skipped = preTurn(turn);

        if (!(turn instanceof TTimeLimit)) {
            try {
                player.inventory.setDeckSlot(player.player.getInventory().getHeldItemSlot(), null, null);
            } catch (ArrayIndexOutOfBoundsException ignored) {}
        }

        if (!skipped && executeTurn(turn)) {
            turn.apply();
            if (turn.damage() != 0.0) {
                if (turn instanceof WinConCategory) {
                    player.piercingDamage(turn.damage());
                } else {
                    player.damage(turn.damage());
                }
            }

            effectManager.applyEffects(turn);
            player.player.playSound(player.player, player.cosmeticPlaceSound(), 0.5f, 1.0f);
            initialPlayers.forEach(p -> p.sendMessage(Translation.component(p.locale(), "gameplay.info.turn-used",
                    player.player.getName(), Utilities.turnName(turn.getClass().getName().replace("net.hectus.neobb.turn.", ""), p.locale())).color(Colors.EXTRA)));
        }

        nextTurn(turn);
        postTurn(turn, skipped);
    }

    public final void nextTurn(@NotNull Turn<?> turn) {
        history.add(turn);
        turn.player().addTurn();
        turnScheduler.tick();

        arena.resetCurrentBlocks();
        resetTurnCountdown();
        players.forEach(p -> p.inventory.sync());

        if (turn.player().hasModifier(Modifiers.P_EXTRA_TURN)) {
            turn.player().removeModifier(Modifiers.P_EXTRA_TURN);
        } else {
            moveToNextPlayer();
        }
        turn.player().validate();

        NeoBB.LOG.info("{}: {} used {}.", id, turn.player().player.getName(), turn.getClass().getName().replace("net.hectus.neobb.turn.", ""));
    }

    /**
     * This is called after the placement and usability were validated.
     * @return {@code true} if the turn should be skipped, {@code false} otherwise.
     */
    public boolean preTurn(Turn<?> turn) { return false; }

    /**
     * This is executed right before {@link Turn#apply()} is called.
     * If the turn is skipped, this will never be executed.
     * @return {@code true} if {@link Turn#apply()} should be called, {@code false} otherwise.
     */
    public boolean executeTurn(Turn<?> turn) { return true; }

    /**
     * This is called after the turn finished.
     * It will be called no matter if the turn got skipped or not.
     * @param skipped If this turn was skipped.
     */
    public void postTurn(Turn<?> turn, boolean skipped) {}

    /**
     * Verifies if the turn can be used and will return true if it is cancelled and the turn method can exit safely, due to being outside of the arena for example.
     * @return {@code false} if the turn method should abort, {@code true} if it can continue.
     */
    public final boolean outOfBounds(@NotNull Location location, Cancellable event) {
        if (Cord.ofLocation(location.clone()).outOfBounds(warp.lowCorner(), warp.highCorner())) {
            event.setCancelled(true);
            return true;
        }
        return false;
    }

    public void outOfBoundsAction(NeoPlayer player) {
        eliminatePlayer(player);
    }

    public @Nullable List<Component> scoreboard(NeoPlayer player) { return null; }
    public @Nullable Component actionbar(NeoPlayer player) { return null; }

    public void gameTick(int tick) {
        if (started) {
            players.forEach(NeoPlayer::tick);
            if (tick % 20 == 0) {
                timeLeft.decrement();
                if (timeLeft.get() == 0)
                    draw(false);
                turnCountdownTick();
            }
        }
    }

    public final List<NeoPlayer> initialPlayers() {
        return initialPlayers;
    }

    public final List<NeoPlayer> players() {
        return players;
    }

    public void eliminatePlayer(@NotNull NeoPlayer player) {
        if (player.hasModifier(Modifiers.P_REVIVE)) {
            player.player.playEffect(EntityEffect.TOTEM_RESURRECT);
            player.sendMessage(Translation.component(player.locale(), "gameplay.info.revive.use").color(Colors.POSITIVE));
            NeoBB.LOG.info("{}: {} used a revive.", id, player.player.getName());
            player.removeModifier(Modifiers.P_REVIVE);
            return;
        }

        players.remove(player);

        player.cosmeticDeathAnimation().play(player);
        player.player.setHealth(0.0);
        NeoBB.LOG.info("{}: {} got eliminated.", id, player.player.getName());

        Bukkit.getScheduler().runTaskLater(NeoBB.PLUGIN, () -> {
            if (players.size() == 1) {
                win(players.getFirst());
            }
            player.player.teleport(warp.center);
            player.player.setGameMode(GameMode.SPECTATOR);
        }, 20);
    }

    public final @NotNull TargetObj gameTarget(boolean onlyAlive) {
        return new TargetObj(onlyAlive ? players : initialPlayers);
    }

    public final List<Turn<?>> history() {
        return history;
    }

    public final World world() {
        return arena.world;
    }

    public final boolean started() {
        return started;
    }

    public final void setStarted(boolean started) {
        this.started = started;
    }

    public final WarpTurn warp() {
        return warp;
    }

    public void warp(@NotNull WarpTurn warp) {
        NeoBB.LOG.info("{}: {} warped from {} to {}.", id, warp.player().player.getName(), this.warp.name, warp.name);

        this.warp = warp;
        allowedClazzes.clear();
        allowedClazzes.addAll(warp.allows());
        modifiers.removeIf(s -> s.startsWith(Modifiers.G_DEFAULT_WARP_PREVENT_PREFIX));

        players.forEach(p -> {
            if (warp.temperature() == WarpTurn.Temperature.COLD) {
                p.player.setFireTicks(0);
                turnScheduler.removeTask("burn");
            }
        });
    }

    public MinecraftTime time() {
        return time;
    }

    public void time(@NotNull MinecraftTime time) {
        this.time = time;
        world().setTime(time.time);
    }

    public boolean allows(Turn<?> turn) {
        for (Class<? extends Clazz> clazz : allowedClazzes) {
            if (clazz.isInstance(turn) && (!(turn instanceof WarpTurn) || !hasModifier(Modifiers.G_DEFAULT_WARP_PREVENT_PREFIX + Utilities.camelToSnake(Utilities.counterFilterName(clazz.getSimpleName())))))
                return true;
        }
        return false;
    }

    public final Set<Class<? extends Clazz>> allowedClazzes() {
        return allowedClazzes;
    }

    public final Time timeLeft() {
        return timeLeft;
    }

    public final void turnCountdownTick() {
        turnCountdown--;
        if (turnCountdown <= 0) {
            if (history.size() >= players.size() && history.subList(history.size() - players.size(), history.size()).stream().allMatch(t -> t instanceof TTimeLimit)) {
                players.forEach(p -> p.sendMessage(Translation.component(p.locale(), "gameplay.info.ending.too-slow").color(Colors.NEUTRAL)));
                draw(false);
            } else {
                turn(new TTimeLimit(new Time(turnCountdown), currentPlayer()), new CancellableImpl());
            }
        }
    }

    public final void resetTurnCountdown() {
        turnCountdown = (int) (info().turnTimer() * difficulty.timeMultiplier);
    }

    public final void moveToNextPlayer() {
        turningIndex = (turningIndex + 1) % players.size();
    }

    public final NeoPlayer getNextPlayer() {
        return players.get((turningIndex + 1) % players.size());
    }

    public final NeoPlayer currentPlayer() {
        if (turningIndex >= players.size())
            turningIndex = 0;
        return players.get(turningIndex);
    }

    public final @Nullable NeoPlayer player(Player player, boolean onlyAlive) {
        for (NeoPlayer p : onlyAlive ? players : initialPlayers) {
            if (p.player == player) return p;
        }
        return null;
    }

    public final boolean allShopDone() {
        for (NeoPlayer player : players) {
            if (!player.inventory.shopDone()) return false;
        }
        return true;
    }

    public void start() {
        timeLeft = new Time((long) (info().totalTime().get() * difficulty.timeMultiplier));
        timeLeft.setAllowNegatives(true);
        resetTurnCountdown();

        NeoBB.LOG.info("{}: Started the game.", id);
        setStarted(true);

        if (info().showIntro()) {
            NeoPlayer current = players.getFirst();
            current.showTitle(Title.title(Component.text("You go first!", Colors.POSITIVE), Component.text("Good Luck!", Colors.NEUTRAL)));
            current.opponents(false).forEach(p -> p.showTitle(Title.title(Component.text(current.player.getName() + " goes first!", Colors.RED), Component.text("Good Luck!", Colors.NEUTRAL))));
        }
    }

    public void end(boolean force) {
        initialPlayers.forEach(p -> cleanPlayer(p.player));
        GameManager.remove(this);
        effectManager.clearHighlight();

        if (force) {
            NeoBB.LOG.info("{}: Stopped the game forcefully.", id);
            return;
        } else {
            NeoBB.LOG.info("{}: Stopped the game gracefully.", id);
        }
        Bukkit.getScheduler().runTaskLater(NeoBB.PLUGIN, () -> {
            initialPlayers.forEach(p -> p.player.kick(Component.text("Game ended!")));
            if (NeoBB.PRODUCTION) {
                Bukkit.getServer().shutdown();
            } else {
                time(MinecraftTime.DAY);
                arena.clear();
            }
        }, 100); // 5 Seconds
    }

    public final void win(@NotNull NeoPlayer player) {
        player.showTitle(Title.title(Translation.component(player.locale(), "gameplay.info.ending.win").color(Colors.POSITIVE),
                Translation.component(player.locale(), "gameplay.info.ending.win-sub").color(Colors.NEUTRAL)));
        player.opponents(false).forEach(p -> p.showTitle(Title.title(Translation.component(p.locale(), "gameplay.info.ending.lose").color(Colors.NEGATIVE),
                Translation.component(p.locale(), "gameplay.info.ending.lose-sub").color(Colors.NEUTRAL))));
        NeoBB.LOG.info("{}: {} won the game.", id, player.player.getName());

        player.cosmeticWinAnimation().play(player);

        end(false);
        if (difficulty.allowRanking)
            Rating.updateRankingsWin(player, player.opponents(false).getFirst(), difficulty.rankingMultiplier); // TODO: Support multiple losers!
    }

    public final void draw(boolean force) {
        players.forEach(p -> p.showTitle(Title.title(Translation.component(p.locale(), "gameplay.info.ending.draw").color(Colors.NEUTRAL),
                Translation.component(p.locale(), "gameplay.info.ending.draw-sub").color(Colors.EXTRA))));
        NeoBB.LOG.info("{}: The game resulted in a draw.", id);

        end(force);
        if (difficulty.allowRanking)
            Rating.updateRankingsDraw(initialPlayers.get(0), initialPlayers.get(1), difficulty.rankingMultiplier); // TODO: Support multiple winners/losers!
    }

    public final void giveUp(@NotNull NeoPlayer player) {
        players.forEach(p -> p.showTitle(Title.title(Translation.component(p.locale(), "gameplay.info.ending.giveup", player.player.getName()).color(Colors.NEUTRAL),
                Translation.component(p.locale(), "gameplay.info.ending.giveup-sub", player.player.getName()).color(Colors.EXTRA))));
        NeoBB.LOG.info("{}: {} gave up.", id, player.player.getName());

        player.opponents(true).forEach(p -> p.cosmeticWinAnimation().play(p));

        end(false);
        if (difficulty.allowRanking)
            Rating.updateRankingsWin(player.opponents(false).getFirst(), player, difficulty.rankingMultiplier); // TODO: Support multiple winners!
    }

    private static void cleanPlayer(@NotNull Player player) {
        player.setScoreboard(Bukkit.getScoreboardManager().getMainScoreboard());
        player.closeInventory();
        player.getInventory().clear();
        player.clearActivePotionEffects();
    }
}

File: src/main/kotlin/net/hectus/neobb/game/Game.kt
package net.hectus.neobb.game

import com.marcpg.libpg.data.modifiable.ModifiableImpl
import com.marcpg.libpg.data.time.Time
import com.marcpg.libpg.display.*
import com.marcpg.libpg.storing.Cord
import com.marcpg.libpg.util.*
import net.hectus.neobb.NeoBB
import net.hectus.neobb.external.cosmetic.EffectManager
import net.hectus.neobb.external.rating.Rating
import net.hectus.neobb.game.util.*
import net.hectus.neobb.matrix.Arena
import net.hectus.neobb.modes.turn.Turn
import net.hectus.neobb.modes.turn.TurnExec
import net.hectus.neobb.modes.turn.default_game.StructureTurn
import net.hectus.neobb.modes.turn.default_game.TDefaultWarp
import net.hectus.neobb.modes.turn.default_game.TTimeLimit
import net.hectus.neobb.modes.turn.default_game.WarpTurn
import net.hectus.neobb.modes.turn.default_game.attribute.TurnClazz
import net.hectus.neobb.modes.turn.person_game.WinConCategory
import net.hectus.neobb.player.NeoPlayer
import net.hectus.neobb.util.*
import net.kyori.adventure.title.Title
import org.bukkit.*
import org.bukkit.entity.BlockDisplay
import org.bukkit.entity.Player
import org.bukkit.event.Cancellable
import kotlin.math.cos
import kotlin.math.sin

abstract class Game(val world: World, private val bukkitPlayers: List<Player>, val difficulty: GameDifficulty = Constants.DEFAULT_DIFFICULTY): ModifiableImpl(), Ticking {
    val id: String = Randomizer.generateRandomString(Constants.GAME_ID_LENGTH, Constants.GAME_ID_CHARSET)
    lateinit var arena: Arena
    val effectManager = EffectManager()
    val turnScheduler = TurnScheduler()

    lateinit var initialPlayers: List<NeoPlayer> private set
    private lateinit var initialTarget: ForwardingMinecraftReceiver
    lateinit var players: MutableList<NeoPlayer> private set
    val history: MutableList<TurnExec<*>> = mutableListOf()

    val allowed: MutableSet<TurnClazz> = mutableSetOf()

    abstract val info: GameInfo

    open val scoreboard: ((NeoPlayer) -> SimpleScoreboard)? = null
    open val actionBar: ((NeoPlayer) -> SimpleActionBar)? = null

    // ==================================================
    // ============ PUBLIC GETTER VARIABLES =============
    // ==================================================

    var warp: WarpTurn = TDefaultWarp
    var playedWarps: List<WarpTurn> = mutableListOf(warp)

    var time: MinecraftTime = MinecraftTime.MIDNIGHT
        set(value) {
            field = value
            world.time = value.time.toLong()
        }

    lateinit var timeLeft: Time
        private set

    var turnCountdown: Int = -1
        private set

    var started: Boolean = false
        private set

    var turningIndex: Int = 0
        private set

    // ==================================================
    // =================== INIT BLOCK ===================
    // ==================================================

    open fun init() {
        Bukkit.getScoreboardManager().mainScoreboard.teams.forEach { it.unregister() }

        initialPlayers = bukkitPlayers.map { NeoPlayer(it, this) }.shuffled()
        initialTarget = initialPlayers.toList().receiver()
        players = initialPlayers.toMutableList()

        arena = Arena(this)

        allowed.addAll(warp.allows)
        resetTurnCountdown()

        val playerCount = players.size
        for (i in 0 until playerCount) {
            val angle = 2 * Math.PI * i / playerCount
            players[i].teleport(
                cord = warp.bounds.center2D + Cord(Constants.SPAWN_POINT_RADIUS * cos(angle), 0.0, Constants.SPAWN_POINT_RADIUS * sin(angle)),
                world = world,
                yaw = Math.toDegrees(angle + Math.PI / 2).toFloat(), // No, I am not a math guy...
                pitch = 0f,
            )
        }

        // Clear any previous block highlights.
        world.getEntitiesByClass(BlockDisplay::class.java).forEach { it.remove() }

        GameManager.add(this)
        info("Initialized the game.")
    }

    // ==================================================
    // ============== ABSTRACT/OPEN MEMBERS =============
    // ==================================================

    /**
     * This is called after the placement and usability were validated.
     * @return `true` if the turn should be skipped, `false` otherwise.
     */
    open fun preTurn(exec: TurnExec<*>): Boolean {
        return false
    }

    /**
     * This is executed right before [Turn.apply] is called.
     * If the turn is skipped, this will never be executed.
     * @return `true` if [Turn.apply] should be called, `false` otherwise.
     */
    open fun executeTurn(exec: TurnExec<*>): Boolean {
        return true
    }

    /**
     * This is called after the turn finished.
     * It will be called no matter if the turn got skipped or not.
     * @param skipped If this turn was skipped.
     */
    open fun postTurn(exec: TurnExec<*>, skipped: Boolean) {}

    /**
     * This is called after the `init {}` block finished.
     */
    open fun extraStart() {}

    /**
     * This is called after the [end] method finished.
     * @param force If the game was stopped forcefully.
     */
    open fun extraEnd(force: Boolean) {}

    /**
     * This is called after the [tick] method finished.
     * @param tick The tick which this method was called for.
     */
    open fun extraTick(tick: Ticking.Tick) {}

    open fun onOutOfBounds(player: NeoPlayer) = eliminate(player)

    open fun allows(turn: Turn<*>): Boolean {
        return turn.clazz == null || turn.clazz == TurnClazz.NEUTRAL || !difficulty.completeRules || turn is WarpTurn || turn.clazz in allowed
    }

    open fun targetPlayer(player: NeoPlayer, beam: Boolean = false): NeoPlayer? = players.following(player)

    // ==================================================
    // ============= CLOSED UTILITY METHODS =============
    // ==================================================

    /**
     * Verifies if the turn can be used and will return `true` if it is canceled and the turn method can exit safely, due to being outside the arena, for example.
     * @return `false` if the turn method should abort, `true` if it can continue.
     */
    fun outOfBounds(cord: Cord, event: Cancellable? = null): Boolean {
        if (cord !in warp.bounds) {
            if (event != null)
                event.isCancelled = true
            return true
        }
        return false
    }

    fun moveToNextPlayer() {
        currentPlayer().player.isGlowing = false
        turningIndex = (turningIndex + 1) % players.size
        currentPlayer().player.isGlowing = true
    }

    fun turnCountdownTick() {
        if (--turnCountdown == 0) {
            if (history.size >= players.size && history.subList(history.size - players.size, history.size).all { it.turn === TTimeLimit }) {
                players.forEach { it.sendMessage(it.locale().component("gameplay.info.ending.too-slow", color = Colors.NEUTRAL)) }
                draw(false)
            } else {
                turn(TurnExec(TTimeLimit, currentPlayer(), null, Time(turnCountdown.toLong())))
            }
        }
    }

    fun resetTurnCountdown() {
        turnCountdown = (info.turnTimer * difficulty.timeFactor).toInt()
    }

    fun nextTurn(exec: TurnExec<*>) {
        history += exec
        exec.player.databaseInfo.addTurn()

        arena.resetCurrentBlocks()
        resetTurnCountdown()
        turnScheduler.tick()

        players.forEach { it.inventory.sync() }

        if (exec.player.hasModifier(Modifiers.Player.EXTRA_TURN)) {
            exec.player.removeModifier(Modifiers.Player.EXTRA_TURN)
        } else {
            moveToNextPlayer()
        }

        if (exec.player.hasModifier(Modifiers.Player.Default.ATTACKED)) {
            eliminate(exec.player)
        }

        info("${exec.player.name()} used ${exec.turn.namespace}.")
    }

    // ==================================================
    // ============== CLOSED BASIC METHODS ==============
    // ==================================================

    fun target(onlyAlive: Boolean = true): MinecraftReceiver = if (onlyAlive) players.receiver() else initialTarget

    fun currentPlayer(): NeoPlayer {
        if (turningIndex >= players.size)
            turningIndex = 0
        return players[turningIndex]
    }

    fun nextPlayer(): NeoPlayer = players[(turningIndex + 1) % players.size]

    fun player(bukkitPlayer: Player, onlyAlive: Boolean = true): NeoPlayer? {
        for (player in (if (onlyAlive) players else initialPlayers)) {
            if (player.uuid() == bukkitPlayer.uniqueId)
                return player
        }
        return null
    }

    fun allShopDone(): Boolean = players.all { it.inventory.shopDone }

    fun info(msg: String) = NeoBB.LOG.info("[${Constants.GAME_LOG_PREFIX}$id] $msg")
    fun warn(msg: String) = NeoBB.LOG.warn("[${Constants.GAME_LOG_PREFIX}$id] $msg")
    fun warn(msg: String, e: Throwable) = NeoBB.LOG.warn("[${Constants.GAME_LOG_PREFIX}$id] $msg", e)

    // ==================================================
    // =============== CLOSED GAME LOGIC ================
    // ==================================================

    override fun tick(tick: Ticking.Tick) {
        if (!started) return

        if (tick.isSecond()) {
            timeLeft--
            if (timeLeft.get() == 0L)
                draw(false)

            turnCountdownTick()
        }

        extraTick(tick)
    }

    fun eliminate(target: MinecraftReceiver) {
        if (target.hasModifier(Modifiers.Player.REVIVE)) {
            target.removeModifier(Modifiers.Player.REVIVE)
            target.eachBukkitPlayer { it.playEffect(EntityEffect.PROTECTED_FROM_DEATH) }

            target.sendMessage("gameplay.info.revive.use", color = Colors.POSITIVE)
            info("${target.name()} has used a revive.")
            return
        }

        target.eachNeoPlayer { p ->
            players.remove(p)
            p.databaseInfo.deathAnimation.play(p)
            p.player.health = 0.0
        }

        bukkitRunLater(Time(1)) {
            if (players.size == 1) {
                win(players.first())
            }
            target.teleport(warp.bounds.center2D)
            target.eachBukkitPlayer { it.gameMode = GameMode.SPECTATOR }
        }

        info("${target.name()} got eliminated.")
    }

    fun <T> turn(exec: TurnExec<T>, event: Cancellable? = null) {
        val player = exec.player
        val nextPlayer = player.targetPlayer()
        val turn = exec.turn

        var allowed = false

        when {
            Constants.CHECK_WARP_CLASSES && !allows(turn) -> {
                notAllowed(exec, "${player.name()} can't use ${turn.name} in the current warp.", "gameplay.info.not_allowed.class", turn.clazz?.name?.toTitleCase() ?: "Unknown", warp.name)
            }
            turn is WarpTurn && hasModifier(Modifiers.Game.NO_WARP.name + "_" + turn.namespace) -> {
                notAllowed(exec, "${player.name()} can't use ${turn.name} right now.", "gameplay.info.not_allowed.warp_filter", turn.name)
            }
            outOfBounds(exec.cord ?: warp.bounds.center3D, event) -> {
                notAllowed(exec, "${player.name()} tried using ${turn.name} out of bounds.", "gameplay.info.out_of_bounds")
            }
            turn.unusable(exec.player) -> {
                notAllowed(exec, "${player.name()} used ${turn.name} incorrectly.", "gameplay.info.wrong_usage")
            }
            else -> allowed = true
        }

        if (turn === TTimeLimit)
            allowed = true

        if (!allowed) return

        if (turn.isCombo) {
            history += exec
            clearSlot(exec)
            return
        }

        val skipped = preTurn(exec)

        clearSlot(exec)

        if (!skipped && executeTurn(exec)) {
            turn.apply(exec)

            if (turn !== TTimeLimit) {
                if (turn.damage != null && turn.damage != 0.0)
                    nextPlayer.damage(turn.damage!!, turn is WinConCategory)

                effectManager.applyEffects(exec)
                initialPlayers.forEach { it.sendMessage("gameplay.info.turn-used", player.name(), turn.translation(it.locale()), color = Colors.EXTRA) }
            }
        }

        nextTurn(exec)
        postTurn(exec, skipped)
    }

    private fun notAllowed(exec: TurnExec<*>, msg: String, key: String, vararg variables: String?) {
        info(msg)
        exec.player.sendMessage(exec.player.locale().component(key, *variables, color = Colors.NEGATIVE))
        exec.player.opponents(false).forEach { it.sendMessage(it.locale().component("gameplay.info.wrong_usage.other", exec.player.name(), color = Colors.NEUTRAL)) }
        exec.player.playSound(Sound.ENTITY_VILLAGER_NO)

        clearSlot(exec)
        nextTurn(exec)
    }

    private fun clearSlot(exec: TurnExec<*>) {
        if (exec.turn !== TTimeLimit) {
            runCatching {
                if (exec.turn is StructureTurn) {
                    exec.turn.referenceStructure.materials.forEach { (m, a) -> exec.player.inventory.clearFirst { i, _ -> i.type == m && i.amount == a } }
                } else {
                    exec.player.inventory.clearSlot(exec.player.player.inventory.heldItemSlot)
                }
            }
        }
    }

    fun warp(player: NeoPlayer, warp: WarpTurn) {
        info("${player.name()} has warped from the ${this.warp.name} to the ${warp.name}.")

        players.forEach { p -> p.sendMessage("gameplay.info.warp.used", this.warp.translation(p.locale()), warp.translation(p.locale()), color = Colors.EXTRA) }

        this.warp = warp
        if (playedWarps.all { it.name != warp.name })
            this.playedWarps += warp

        allowed.clear()
        allowed.addAll(warp.allows)
        modifiers.removeIf { it is String && it.startsWith(Modifiers.Game.NO_WARP.name) }

        players.forEach { p ->
            p.removeModifier(Modifiers.Player.Default.DEFENDED)
            if (warp.temperature == WarpTurn.Temperature.COLD) {
                p.player.fireTicks = 0
                turnScheduler.remove(ScheduleID.BURN)
            }

            if (difficulty.completeRules) {
                p.sendMessage("gameplay.info.warp.playable", this.warp.translation(p.locale()), warp.translation(p.locale()), color = Colors.NEUTRAL)
                warp.allows.forEach { p.sendMessage(it.line(p.locale())) }
            }
        }
    }

    fun start() {
        timeLeft = Time((info.totalTime.get() * difficulty.timeFactor).toLong())
        timeLeft.allowNegatives = true
        resetTurnCountdown()

        info("Game has been started.")
        started = true

        players.forEach { it.start() }

        if (info.showIntro) {
            val currentPlayer = players.first()
            currentPlayer.showTitle(Title.title(currentPlayer.locale().component("gameplay.info.start.first.you", color = Colors.POSITIVE),
                currentPlayer.locale().component("gameplay.info.start.first-sub", color = Colors.POSITIVE)))
            currentPlayer.opponents(false).forEach { it.showTitle(Title.title(it.locale().component("gameplay.info.start.first", currentPlayer.name(), color = Colors.POSITIVE),
                it.locale().component("gameplay.info.start.first-sub", color = Colors.POSITIVE))) }
        }

        currentPlayer().player.isGlowing = true

        extraStart()

        bukkitRunTimer(20, 20) {
            if (id !in GameManager.GAMES) {
                it.cancel()
            } else {
                if (!players.contains(currentPlayer()))
                    moveToNextPlayer()
            }
        }
    }

    fun end(force: Boolean = false) {
        initialPlayers.forEach {
            it.clear(display = true)
            it.team.unregister()
        }
        GameManager.remove(this)
        effectManager.clearHighlight()

        if (force) {
            warn("Stopped the game forcefully.")
            extraEnd(true)
            return
        }

        bukkitRunLater(Time(Constants.GAME_CLEANUP_DELAY)) {
            if (Configuration.PRODUCTION) {
                target(false).sendMessage("gameplay.info.kick", color = Colors.NEUTRAL)
                Bukkit.getServer().shutdown()
            } else {
                time = MinecraftTime.DAY
                arena.clear()
            }
        }

        extraEnd(false)
        info("Stopped the game gracefully.")

        // Save replay if gracefully stopping the game.
        runCatching {
            saveReplay()
        }.onFailure {
            warn("Unable to save replay due to internal error: ${it.message}")
        }
    }

    fun win(player: NeoPlayer) {
        player.showTitle(Title.title(player.locale().component("gameplay.info.ending.win", color = Colors.POSITIVE),
            player.locale().component("gameplay.info.ending.win-sub", color = Colors.NEUTRAL)))
        player.opponents(false).forEach { it.showTitle(Title.title(it.locale().component("gameplay.info.ending.lose", color = Colors.NEGATIVE),
            it.locale().component("gameplay.info.ending.lose-sub", color = Colors.NEUTRAL))) }
        info("${player.name()} has won the game.")

        player.databaseInfo.winAnimation.play(player)

        end(false)
        if (difficulty.ranked)
            Rating.updateFFARankings(initialPlayers.map { it.databaseInfo }, player.databaseInfo)
    }

    fun draw(force: Boolean = false) {
        players.forEach { it.showTitle(Title.title(it.locale().component("gameplay.info.ending.draw", color = Colors.NEUTRAL),
            it.locale().component("gameplay.info.ending.draw-sub", color = Colors.EXTRA))) }
        info("The game resulted in a draw.")

        end(force)
        if (difficulty.ranked)
            Rating.updateFFARankings(initialPlayers.map { it.databaseInfo }, null)
    }
}
